src/server/snapshot.cc

```diff
@@ -235,30 +239,35 @@ void SliceSnapshot::IterateBucketsFb(const Cancellation* cll, bool send_full_syn
 }
 
 bool SliceSnapshot::BucketSaveCb(PrimeIterator it) {
+  BucketSerializationGuard guard(this);
+
   ++stats_.savecb_calls;
 
+  auto check = [&](auto v) {
+    if (v >= snapshot_version_) {
+      // either has been already serialized or added after snapshotting started.
+      DVLOG(3) << "Skipped " << it.segment_id() << ":" << it.bucket_id() << ":" << it.slot_id()
+               << " at " << v;
+      ++stats_.skipped;
+      return false;
+    }
+    return true;
+  };
+
   uint64_t v = it.GetVersion();
-  if (v >= snapshot_version_) {
-    // either has been already serialized or added after snapshotting started.
-    DVLOG(3) << "Skipped " << it.segment_id() << ":" << it.bucket_id() << ":" << it.slot_id()
-             << " at " << v;
-    ++stats_.skipped;
+  if (!check(v)) {
     return false;
   }
+
   db_slice_->FlushChangeToEarlierCallbacks(current_db_, DbSlice::Iterator::FromPrime(it),
```

Comment:
I am trying to understand how this works as FlushChangeToEarlierCallbacks will also call serialize bucket for snapshot with earlier version and they will be locked on the mutex/BucketSerializationGuard . is that right?

Reply:
This should not be the case because the callbacks we apply are registered via a different `slice_snapshot` and their `guard` is a different instance. Furthermore, the registered callback of the sliced snapshot instance that called `FlushChangeToEarlierCallbacks` won't be called because of this condition:

```
  1120     if (cb_version == upper_bound) {                                                              
  1121       return;                                                                                     
  1122     }  
```

And the `upper_bound` is the version that we got when we registered the callback in `SliceSnapshot::Start`. So we flush all the previous callbacks excluding the one registered by this instance.